wordpress-mobile/WordPress-iOS
# [JP Plugin] Move Remote Install tracking logic to the view model

Refs #20018
Depends on #20066 👈🏼 Please review this one first.

The Jetpack plugin installation (through the Jetpack Proxy) will require different tracking events. This PR moves the tracking responsibility to the view model, and the view model class is abstracted so a different instance can be plugged into the view controller.

Apart from the tracking change, here are some other changes included in this PR:

- `JetpackRemoteInstallViewModel` has now become a protocol.
- The previous concrete implementation is now renamed into `SelfHostedJetpackRemoteInstallViewModel`.
- The `installJetpack` method now only requires a `Blog` instance, instead of previously requiring the username or password.
  - This is done to later accommodate the REST API installation method; so the required parameters are now decided from the view model.
- Cleaned up some unused properties.

## To test

Since this is only a (relatively small) refactor work, please verify that the existing Tracking implementation is not impacted:

- Create a self-hosted site, and have the Jetpack plugin deactivated and uninstalled.
- Log into the site through the site address method (username & password).
- Navigate to the Notifications tab, and tap "Install Jetpack".
- Turn on airplane mode on your device (so we test the actual installation completion last).
- Tap "Continue", and verify that these events are fired:
  - `🔵 Tracked: install_jetpack_remote_start <>`
  - `🔵 Tracked: install_jetpack_remote_failed <error_type: unknown, site_url: your-site-url>`
- Tap "Retry", and verify that these events are fired:
  - `🔵 Tracked: install_jetpack_remote_restart <>`
  - `🔵 Tracked: install_jetpack_remote_failed <error_type: unknown, site_url: your-site-url>`
- Turn off airplane mode.
- Tap "Retry" again, and verify that these events are fired:
  - `🔵 Tracked: install_jetpack_remote_restart <>`
  - `🔵 Tracked: install_jetpack_remote_completed <>`
- Tap "Set up", and verify that this event is fired:
  - `🔵 Tracked: install_jetpack_remote_connect <>`

> **Note**: There's no need to test further than this, as it's out of the scope of this project. Also worth noting, there are known issues with the Jetpack connection setup process (which will not be addressed here). For reference, @staskus has compiled these known issues in #19261.

## Regression Notes
1. Potential unintended areas of impact
Possibly impact the tracking behavior of the existing Jetpack Remote Install flow.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes.

3. What automated tests I added (or what prevented me from doing so)
N/A.

PR submission checklist:
